collab-document/tests/document/redo_undo_test.rs

```diff
@@ -0,0 +1,238 @@
+use collab_document::blocks::Block;
+use nanoid::nanoid;
+use serde_json::to_value;
+use std::collections::HashMap;
+use std::time::Duration;
+
+use crate::util::{
+  create_document, create_document_with_db, db, delete_block, get_document_data, insert_block,
+  open_document_with_db, update_block,
+};
+
+#[tokio::test]
+async fn insert_undo_redo() {
+  let doc_id = "1";
+  let test = create_document(1, doc_id);
+
+  tokio::time::sleep(Duration::from_secs(1)).await;
+
+  let (page_id, _, _) = get_document_data(&test.document);
+  let block_id = nanoid!(10);
+  let block = Block {
+    id: block_id.clone(),
+    ty: "paragraph".to_string(),
+    parent: page_id.to_string(), // empty parent id
+    children: "".to_string(),
+    external_id: None,
+    external_type: None,
+    data: Default::default(),
+  };
+
+  insert_block(&test.document, block.clone(), "".to_string()).unwrap();
+
+  let can_undo = test.document.can_undo();
+  assert!(can_undo);
+  let undo = &test.document.undo();
+  assert_eq!(undo.to_owned(), true);
+  let insert_block = &test.document.get_block(&block_id);
+  assert_eq!(insert_block.is_none(), true);
+
+  let can_redo = test.document.can_redo();
+  assert!(can_redo);
+  let redo = &test.document.redo();
+  assert_eq!(redo.to_owned(), true);
+  let insert_block = &test.document.get_block(&block_id);
```

Comment:
Most of the case, you don't need the using &. get_xxx should return reference or an object.